initial work for basisu

cdhanna opened the pull request:
This is meant to be public for transparency, but I'm not ready to open it to the Monogame repo yet. 

As of Aug 15th, here is what I think I have left to do...
- [ ] formalize `basisu` distribution with monogame
- [x] find a better way to save an image file before it gets input to `basisu` CLI. Maybe there is even a way to pipe the input
- [x] formalize ATC
- [x] formalize ETC 
- [x] formalize Pvrtc
- [x] figure out what to do about DXT1a not being supported by `basisu`. 
- [x] figure out what to do about DXT3 not being supported by `basisu`
- [x] write unit tests for other formats if possible
- [x] figure out how to log possible errors or warnings, or commit to _not_ doing this
- [x] add support for ETC2 and ATSC
- [ ] test

chrisbeamable committed:
initial work for basisu

chrisbeamable committed:
removes comments and ingests libKTX nuget

chrisbeamable committed:
adds tests for etc1 atc and pvrtc

dellis1972 commented:
So according to Tom DXT3 maps to BC5 not BC3. So it might be worth seeing what DXT formats map to the newer BC ones. It might be we are covered as its not a direct mapping.

cdhanna commented:
[Wikipedia](https://en.wikipedia.org/wiki/S3_Texture_Compression#BC4_and_BC5) and [Microsoft](https://learn.microsoft.com/en-us/windows/win32/direct3d10/d3d10-graphics-programming-guide-resources-block-compression#bc2) suggest that DXT3 is part of BC2. `basisu` does not support BC2

Those docs also suggest that DXT5 is BC3. 
DXT5 is part of BC3, which I've specified, https://github.com/cdhanna/MonoGame/blob/7b35042fad4a8de9a5079967a21b02a497acc19f/MonoGame.Framework.Content.Pipeline/Utilities/BasisUHelpers.cs#L117

`basisu` doesn't support alpha punchthrough on DXT1, which means `DXT1a` is not supported. 
https://github.com/BinomialLLC/basis_universal/blob/master/transcoder/basisu_transcoder.h#L58

In summary, I believe that `basisu` only supports
- DXT1
- DXT5

Which means that the following are not supported, 
- DXT1a
- DXT3

Because the content tool isn't actually exposing the exact DXT version, but rather, "DXT compression", we _could_ converge those unsupported formats onto DXT5. I know someone said that there is a custom DXT3 compressor for spritefonts, but it seems to be greyscale only. 



----

Separately, I've noticed that `basisu` will write the resulting compression `.ktx` files successfully for, 
- DXT1
- DXT5
- ETC1
- PVRTC

But when I ask `basis` to unpack the file for ATC_INTERPOLATED, it just doesn't write a `.ktx` file, and there appear to be no errors on stdout or stderr. I'm still investigating and reading through the tool and maybe I'm missing some required setup for ATC specifically. Anyway, I just wanted to communicate that part. 

And on that note, `AtcRgbaExplicitAlpha` is not supported by `basisu` either :( 


----

Also, it stinks that `basisu` _requires_ file-io to do anything. It would be fantastic if we could pipe the uncompressed data into the tool, and the compressed data out of the tool. However, it requires that files be used instead, and I'm struggling to find a compelling place to put those intermediate files. The existing architecture of the `BitmapContent` class makes it a bit awkward to get access to the original content's intermediate directory. I _could_ go plump that data through the system, but I'd like to find a better solution. I've also entertained the idea in my head to just write/read the files from the system's temp directory, and delete the files in a `finally` block. That also leaves a sour taste in my mouth. 

I know there is the IntermediateDirectory on the context, 
```csharp
context.IntermediateDirectory
```

So maybe I just need to suck it up and refactor the `BitmapContext` accept this intermediate directory. 

----

Finally, (for now, haha), I see that there is a logger available in mgcb, but similarly to the intermediate file location, it isn't being provided into the `BitmapContent` types, and I'm hesitant to add it, because it feels like creep. On the other hand, I find myself longing for the ability to emit some verbose logs. Maybe I'll do some investigating how STDOUT / STDERR are handled in mgcb. I could just `Console.Writeline` something, or perhaps there are statically available loggers somewhere. I could put them behind a compiler symbol like `#if ENABLE_VERBOSE_BASISU_LOGS`... 🤔

chrisbeamable committed:
atc is always interpolated, and dxt3 becomes dxt5

cdhanna commented:
I will commit tomorrow probably, but I've decided to plumb the context through the Bitmap content classes. this gives access to the logger and the intermediate path, which are the two things I wanted. 

I'm patching up the tests now from the fallout

cdhanna commented:
well I bit the bullet and decided to plumb the context through the entire `BitmapContext` architecture. Tests are passing again, and now I have the added benefit of being able to use the intermediate directory to create the input `png`, intermediate file, and compressed `ktx` files. I also have access to the logger, which is nice, so I can log any failure messages. 

----

I've been reading up on ETC2 support and I think I'll tackle that (and the ATSC) support next.

dellis1972 commented on MonoGame.Framework.Content.Pipeline/IContentContext.cs:
```diff
@@ -0,0 +1,14 @@
+namespace Microsoft.Xna.Framework.Content.Pipeline;
```
If this in a non XNA class it should probably be in a MonoGame namespace, especially as it adds to the public api.

dellis1972 commented on MonoGame.Framework.Content.Pipeline/Graphics/PixelBitmapContent.cs:
```diff
@@ -17,7 +17,7 @@ public class PixelBitmapContent<T> : BitmapContent where T : struct, IEquatable<
         internal SurfaceFormat _format;
 
-        public PixelBitmapContent(int width, int height)
+        public PixelBitmapContent(int width, int height, IContentContext context) : base(context)
```
Adding these additional arguments will break the public API compatibility with XNA (see https://learn.microsoft.com/en-us/previous-versions/windows/xna/bb195980(v=xnagamestudio.10)). 

We might need a different way to get the "current" context, perhaps a static factory class or something similar?

cdhanna replied:
Oh man, I completely spaced and forgot about api compatibility. _Whoops_. Thanks for check! I'll set about coming up with alternatives. I like the suggestions you gave.

dellis1972 commented:
Generally errors are of type `PipelineException` which are caught here https://github.com/MonoGame/MonoGame/blob/develop/Tools/MonoGame.Content.Builder/BuildContent.cs#L406

For logging warnings/errors you might be able to use something like this https://github.com/MonoGame/MonoGame/blob/develop/MonoGame.Framework.Content.Pipeline/Processors/EffectProcessor.cs#L103.

dellis1972 commented:
I just tried ATC (format 10) on a test file and it worked. 

`basisu -unpack foo.ktx2 -ktx_only -linear -format_only 10`

```
Transcode of layer 0 level 0 face 0 res 64x64 format ASTC_RGBA succeeded
Wrote KTX file "foo_transcoded_ASTC_RGBA_0000.ktx"
Success

```

I wonder if it has to do with the source file?

dellis1972 commented on MonoGame.Framework.Content.Pipeline/MonoGame.Framework.Content.Pipeline.csproj:
```diff
@@ -86,40 +77,18 @@
     <PackageReference Include="Microsoft.NETCore.App" Version="2.1.30" />
     <PackageReference Include="SharpDX" Version="4.0.1" />
     <PackageReference Include="SharpDX.D3DCompiler" Version="4.0.1" />
+    <PackageReference Include="LibKTX" Version="0.9.2" />
```
ideally we'd like to have something similar to the `DdsLoader` for the ktx format. But this is good for testing purposes 😄 I'll see if I can dig up the one I started but never had time to finish.

cdhanna replied:
Not sure I agree with you on this; the LibKTX is the nuget package for `libktxsharp`, referenced in the Monogame fork of basisU, https://github.com/MonoGame/MonoGame.Tool.BasisUniversal. 

That dll just wasn't available on nuget, until (very strangely timed), I opened a PR to do so, and the maintainer also had a commit ready to rock. https://github.com/mcraiha/libktxsharp/pull/5 

Why would we want to pull in the library without Nuget?

dellis1972 replied:
In this case we only need to ability to read ktx files, so pulling in a full library might be a bit overkill. 
We can leave this as is for now and see what the other maintainers think when we open the PR on the main repo.

dellis1972 replied:
The library was mentioned in the BasisU fork because it was an example that I used to test it worked.

dellis1972 commented:
Btw there is an example "API" wrapper for webgl here https://github.com/BinomialLLC/basis_universal/blob/master/webgl/transcoder/basis_wrappers.cpp. We might be able to create a dll based on that to do the job. That said though if we decide to switch to using a dll/so/dylib rather than an exeutable it will mean changing all the MonoGame.Tool.BasisUniversal over to be a Library.

chrisbeamable committed:
revert changes that break namespaces

chrisbeamable committed:
use static factory to handle context operations

cdhanna commented:
> We might need a different way to get the "current" context, perhaps a static factory class or something similar?

Maybe I overcomplicated it a bit, but I just added a class, https://github.com/cdhanna/MonoGame/pull/1/commits/e8a9d0080eb3a51ddcfab4f4752c40a9892eca85#diff-470ae930051bf2b7c89a33462b3942b71f1cd03530cf9c1d62eac63b8912ed6a , that acts a static singleton that holds onto the various contexts. I still needed to modify the innards of the xna content classes, but the type&method signatures are unchanged.

cdhanna commented:
> I just tried ATC (format 10) on a test file and it worked.
> 
> `basisu -unpack foo.ktx2 -ktx_only -linear -format_only 10`
> 
> ```
> Transcode of layer 0 level 0 face 0 res 64x64 format ASTC_RGBA succeeded
> Wrote KTX file "foo_transcoded_ASTC_RGBA_0000.ktx"
> Success
> ```
> 
> I wonder if it has to do with the source file?

Hey, sorry for the delay on this one---
You said you used `ATC` format as code 10, but according to basisU's docs, code 10 maps to ASTC. 
https://github.com/BinomialLLC/basis_universal/blob/master/transcoder/basisu_transcoder.h#L69

ATC itself has formats 11 and 12, which map to no alpha and interpolated alpha. I was saying that ATC interpolated alpha (code 12) does not generate a file from basis. 

From my (albeit limited research), ASTC is not a direct replacement for ATC, but seems _newer_ (and maybe just generally all around _better_) than ATC. So in that sense; perhaps the user's selection for ATC could forward map to ASTC? 

----

I'm also still quite uncertain how to handle the comments about DXT https://github.com/cdhanna/MonoGame/pull/1#issuecomment-2261749298, 

The only thing I can think of that seems within reasonable scope is to make DXT1a and DXT3 just get sneakily turned into DXT5 ... which feels plain silly...

cdhanna commented:
The work-order in the original write up states, 
>Add Support for the following Texture Compression Formats
Etc2
Atsc

As far as modifying XNA namespaces, the `TextureProcessorOutputFormat` type is in an XNA namespace, `Microsoft.Xna.Framework.Content.Pipeline.Processors`. 

Since this work is just _adding_ enum values, do you think it is okay to modify the enum type? Or would ETC2 support get snuck into ETC1 selection, and ASTC get snuck into ATC? 

---

And I guess the same question in sprit applies to creating new classes for `Ect2Bitmap` and `AstcBitmap`; for consistency, it feels like those should be _added_ to the XNA namespace; but I'm equally happy plopping those new types into a new Monogame namespace.

chrisbeamable committed:
uses basisu tool

dellis1972 commented:
> From my (albeit limited research), ASTC is not a direct replacement for ATC, but seems _newer_ (and maybe just generally all around _better_) than ATC. So in that sense; perhaps the user's selection for ATC could forward map to ASTC?

Sorry  you are right, ASTC is not the same as ATC, my bad. I however just tested formats 11 and 12 and they work here. So it might be related to the source image size perhaps (power of 2), although I didn't think it had any restrictions. 

> I'm also still quite uncertain how to handle the comments about DXT [#1 (comment)](https://github.com/cdhanna/MonoGame/pull/1#issuecomment-2261749298),
> 
> The only thing I can think of that seems within reasonable scope is to make DXT1a and DXT3 just get sneakily turned into DXT5 ... which feels plain silly...

I'm asking @TomSpilman for some input on this as I don't know a huge amount about the differences between these formats. I remember tom saying the DXT3 maps to BC5 in the new format system.

dellis1972 commented:
> The work-order in the original write up states,
> 
> > Add Support for the following Texture Compression Formats
> > Etc2
> > Atsc
> 
> As far as modifying XNA namespaces, the `TextureProcessorOutputFormat` type is in an XNA namespace, `Microsoft.Xna.Framework.Content.Pipeline.Processors`.
> 
> Since this work is just _adding_ enum values, do you think it is okay to modify the enum type? Or would ETC2 support get snuck into ETC1 selection, and ASTC get snuck into ATC?
> 
> And I guess the same question in sprit applies to creating new classes for `Ect2Bitmap` and `AstcBitmap`; for consistency, it feels like those should be _added_ to the XNA namespace; but I'm equally happy plopping those new types into a new Monogame namespace.

There new enums can be added to the existing enum type as the new classes can also do in the XNA namespace. The original XNA did not have a PVRTCBitmap or a ETCBitmap or the associated enums as it was based on DirectX only. 

It was just any helper classes outside of the XNA api should go into a MG namespace... I know its not consistent but it's what we have been doing.

ThomasFOG commented:
For DXT, I believe that we're good with ditching DXT3 and just assuming no alpha = DXT1, alpha cutout = DXT1a, alpha range = DXT5.

And yeah, DXT1/3/5 are BC1/2/3.

For PC games, the industry standard is toward "crunched DXT1/5", because it's the most versatile combo and widely supported format. DXT3 is almost never used because it is not optimal, which is why most modern compressors don't bother implementing it.

I don't think that it is worth keeping ATC.
It's likely worth supporting ASTC, even though ETC still is the most commonly supported format on mobile.

Same thing for BC4+, those are too specific to be interesting for general use.

dellis1972 commented:
Well since its an interim release, dropping ATC would not be a good move. ATC is still used on mobile (Android) along with PVRTC.  ETC1 does not support Alpha. 

We use our own DXT3 texture compression code (C#) for fonts , that should stay as changing the font compression could break the look of text in games.

dellis1972 commented:
on a side note it does look like ATITC (ATC) does have a multiple of 4 size restriction. So similar to PVRTC in that case.

dellis1972 commented:
For DXT1a I guess since we can't support alpha punch through if an alpha channel is present has transparency, we should map that to BC3.

tomspilman commented:
We will have to add another compressor to support DXT1a and DXT3.

I say keep things simple and use STB:

tomspilman commented:
We will have to add another compressor to support DXT1a (6x compression) and ~~DXT3 (sharp non-premultiplied alpha at 4x compression used in font rendering)~~.

I say keep things simple and use STB for these cases specifically:

https://github.com/nothings/stb/blob/master/stb_dxt.h

Supports these formats, super portable, and is easy to work with.  We just need a native wrapper for it which is always the complexity.

> I've also entertained the idea in my head to just write/read the files from the system's temp directory, and delete the files in a finally block.

@cdhanna 

This is how we do things in the console ports when we have to run 3rd-party tools like shader compilers that want to work via the command line and generate files.

Just use `Path.GetTempFileName()` and then `ExternalTool.DeleteFile(xxxx)` in a `finally` block.  It works fine.

Yeah it feels wrong, but you have no choice if we want to use these tools.

UPDATE:

So we don't need to support DXT3 really... because we have a C# level DXT3 compressor already specifically for fonts.

dellis1972 commented:
There is also this compressor https://github.com/illwieckz/crunch. But as STB does this already we should use that. We already use Stb via https://github.com/StbSharp/StbImageSharp/. Which I think its the same library?

dellis1972 commented:
Actually ignore me. StbImageSharp only exposes the image stuff.

dellis1972 commented:
There is a library for StbDXT for C# https://github.com/StbSharp/StbDxtSharp

tomspilman commented:
> There is also this compressor https://github.com/illwieckz/crunch.

It would be good to support Crunch as an option.

If it can do a pure DXT1a compressed texture without "crunch"-ing then it is a good fit.

We can then in addition support Crunch textures.  Wwe would need to add a "quality" setting of some sort to the texture options. 
 But in addition it requires runtime support as textures need to be decoded back from crunch to a GPU friendly format on load.

So still something we want to do, but maybe not for this PR.

tomspilman commented:
> There is a library for StbDXT for C# https://github.com/StbSharp/StbDxtSharp

As long as we keep this to the pipeline then maybe performance is acceptable.

These C# "ports" of STB header libs are often much slower than the originals, but it isn't so much of a problem in the content pipeline.

Our big win in the content pipeline is processing multiple content items in parallel.  So pure performance on a file by file basis isn't as critical IMO.

cdhanna commented:
It doesn't seem like stb_dxt supports dxt1a _either_. It does support dxt1, but not dxt1a. 
https://github.com/GammaUNC/FasTC/issues/26

It also looks like [https://github.com/illwieckz/crunch](https://github.com/illwieckz/crunch) states it doesn't support dxt1a either, 
>The CRN file format supports the following core DXTn texture formats: DXT1 (but not DXT1A), DXT5, DXT5A, and DXN/3DC

I took a look at the source code for stb_dxt to see if I could add support for the [degeneracy](https://www.reedbeta.com/blog/understanding-bcn-texture-compression-formats/#degeneracy-and-breaking-it) to support dxt1a, but I'm afraid that is a bit beyond my skill at the moment. 


----

However, maybe it is worth circling back to the start of this whole convo, and revisit what the content processor actually does today, 

The `CompressDxt` [method](https://github.com/cdhanna/MonoGame/blob/develop/MonoGame.Framework.Content.Pipeline/Graphics/GraphicsUtil.cs#L156) has this snippet, 
```csharp
            if (isSpriteFont)
                CompressFontDXT3(content);
            else if (alphaRange == AlphaRange.Opaque)
                content.ConvertBitmapType(typeof(Dxt1BitmapContent));
            else if (alphaRange == AlphaRange.Cutout)
                content.ConvertBitmapType(typeof(Dxt3BitmapContent));
            else
                content.ConvertBitmapType(typeof(Dxt5BitmapContent));
```

The custom DXT3 compressor is only used when the asset is a sprite font. That function has a comment mentioning it is only designed for greyscale font textures; so I don't think it would be generalizable to color images. 

```csharp
        // Compress the greyscale font texture page using a specially-formulated DXT3 mode
        static public unsafe void CompressFontDXT3(TextureContent content)
```

I did try coercing a bitmap through that function anyway, but the resulting compressed bytes do not come through the DXT3 decompressor correctly in the unit tests. (which makes sense, because it seems like the font method is doing something specialized to fonts)
```csharp
                    var tc = new Texture2DContent();
                    tc.Faces[0] = sourceBitmap;
                    GraphicsUtil.CompressFontDXT3(tc);
                    compressedBytes = tc.Faces[0][0].GetPixelData();
```

So in this regard, I do not think we can use the existing DXT3 font compressor for non font images. 

----

The `Dxt3BitmapContent` is only generated if `(alphaRange == AlphaRange.Cutout)`. The `Cutout` property is only used when the only alpha values in the image `0` or `1`. 
I would suggest that we change the `CompressDxt` method to use DXT5 when alphaRange is `Cutout`, and then no DXT3 is required. 

----

Next, the `Dxt1BitmapContent` is only generated if the `alphaRange == AlphaRange.Opaque)`
The `Opaque` value is only selected if all of the alpha values in the image are `1`. 

Internal to the `DxtBitmapContent`, there is this check, 
```csharp
            switch (format)
            {
                case SurfaceFormat.Dxt1:
                case SurfaceFormat.Dxt1SRgb:
                {
                    bool hasTransparency;
                    PrepareNVTT_DXT1(sourceData, out hasTransparency);
                    outputFormat = hasTransparency ? Format.DXT1a : Format.DXT1;
                    alphaMode = hasTransparency ? AlphaMode.Transparency : AlphaMode.None;
                    alphaDither = true;
                    break;
                }
```

`DXT1a` is only selected if the `hasTransparency` variable is true; but it _won't_ be, since we've already decided that the DXT1 was selected given alphaRange was `Opaque`. Therefor, I think this whole saga about DXT1a not being supported _doesn't_ matter. The only other place the `Dxt1BitmapContent` class is used is from the `DdsLoader`, and it only checks for the surface format, Dxt1. 

I don't know what the expectation is around the `DxtBitmapContent` but it seems like its only used within the Content Pipeline project. I'm just trying to say that in the code that calls `DxtBitmapContent` in the pipeline, dxt1a is not currently possible _anyway_, because if there was alpha, it would have been assigned to dxt3. 

----



So long story short, I think the biggest issue is _actually_ just to map the alphaRange Cutout option to DXT5. Dxt1a was _never actually used_, so its a red-herring. But also, thats good, because it seems none of the current tools support it anyway.

dellis1972 commented:
@cdhanna 

Thats good info, never realised that we only called the Dxt1 compression for Opaque textures. If that is the case I wont why the dxt1 tests failed? because it would suggest that basisu bc1 support should be ok for opaque textures. If they are failing then we can probably use stbdxt for the dxt1 path?

dxt3 we will need some input from @tomspilman on that one.

cdhanna commented:
the dxt1 tests fail because there is a place where it compresses a texture with alpha. So I guess I was slightly misleading before, the Test used dxt1a

dellis1972 commented:
I'm working on getting this working https://github.com/MonoGame/MonoGame.Tool.Crunch as a backup.

chrisbeamable committed:
fix

chrisbeamable committed:
comment

cdhanna commented:
Okay, I've pushed some changes that I _think_ put this PR within the goalposts for submitting to the main Monogame repo. 

I found the issue with the ATC bump I ran into last week. It turns out that if you provide the `-uastc` flag while generating the intermediate ktx file, transcoding the intermediate file to ATC doesn't work. But, mercifully, if you omit the `-uastc` flag, ATC transcoding is successful. I've extended my new `BasisUFormat` struct to allow formats to specify if they want to include the `-uastc` flag, and so far, ONLY the ATC format ignores it. 

I made some explicit `NotSupportedExceptions` get tossed if the `DxtBitmapContent` is used when either, 
1. the format is DXT1 and there pixels with non opaque alpha detected, or
2. the format is DXT1a regardless of actual pixel content, or
3. the format is DXT3 regardless of actual pixel content 

I co-opted the existing `TextureProcessorOutputFormat.Etc1Compressed` enum to actually use ETC2 compression if the pixel data contain non-opaque alpha. When these pixels were detected previously, it would fallback to `PixelBitmapContent<Bgra4444>`. ETC1 is still used if there are no alpha pixels. 

I added a `TextureProcessorOutputFormat.AstcCompressed` enum that jumps straight to the newly added `AstcBitmapContent` type which uses the newly added `SurfaceFormat.ASTC_4x4_Rgba`, which in turn maps to the basisU format `cTFASTC_4x4_RGBA` (code 10). 

I've added some loose tests for ETC1, ETC2, ATC, ASTC, PVR, but they only test _compression_; I don't have anything in place to _decompress_ them back into a testable RGBA format like the DXT tests have. 
I modified the DXT tests to _assert_ a `NotSuportedExpception` is thrown in the DXT exception cases described above. 


-----


At this point, my biggest TODO has to do with actually _testing_ this full scale. On the one hand, I think I could fiddle around enough to get this running from source on my windows machine and give that a whirl. But as for the mac testing, because the `FreeImage` library is not ported yet, I still can't pass all the unit tests anyway, so I'm doubtful the build would work all the way. But perhaps it would. 

Another large task that needs to happen is the official publication of the `mgcb-basisu` tool package. For my own testing, I forked the tool repo, tagged it, and used the private github nuget source to install the tool locally. However, for release, I think we'd need to...
1. tag an official release of the mgcb-basisu tool, 
2. upload the package to nuget.org (or another publicly accessible nuget source that doesn't require custom authentication),
3. coerce the `.config/dotnet-tools.json` file to include the right version of the `mgcb-basisu` tool.

chrisbeamable committed:
clean files instead of leaving them in intermediate

dellis1972 commented:
This is all great work thank you. :)

We might be able to use crunch for DXT3? I'm not sure how much work it would be though. It looks like it does support it providing you don't use a CRN output format or certain DDS... But it does support ktx. And I just completed this https://github.com/MonoGame/MonoGame.Tool.Crunch. (Although it does need publishing ).

```
DXT3 is not supported when writing CRN or clustered DXTn DDS files. (DXT3 is supported by crnlib's when compressing to regular DXTn DDS files.) You'll get DXT5 files if you request DXT3. However, DXT3 is supported by the regular DXTn block compressor. (DXT3's 4bpp fixed alpha sucks verses DXT5 alpha blocks, so I don't see this as a bug deal.)
```
I'd rather not drop a format unless we can really help it. Especially as this will be going out in a non major release. 

I wonder if we need to separate `TextureProcessorOutputFormat.EtcCompressed` which will do both ETC1 and ETC2 depending on the Alpha and document the behaviour in the xml comments. And deprecate the Etc1Compressed (somehow).

cdhanna commented:
This is great! 
I forked and made a private [release](https://github.com/cdhanna/MonoGame.Tool.Crunch/pkgs/nuget/mgcb-crunch) for the mgcb-crunch tool, and integrated it quick and dirty.

And DXT3 works! And DXT1A works too! The documentation I found must have been wrong, because when I ran the `-h` flag, I saw clear as day the support for `DXT1A`, and sure enough, when I plugged it in, the old unit tests started passing. 
I'll need a bit more time to clean this up, but things are looking great!

To give you a sense of the command I'm passing to Crunch, 
```
$"-file {pngFileName} -DXT3 -forceprimaryencoding -noNormalDetection -compressor RYG -out {intermediateFileName} -fileformat ktx"
```

Seems like it works if I pass the regular CRN compressor too; although the docs made it seem as though that shouldn't work 🤔 I must have misunderstood them.

chrisbeamable committed:
crunch support for dxt1a and dxt3, use crunch for all dxt

chrisbeamable committed:
deprecate Etc1Compressed ; introduce EtcCompressed

dellis1972 reviewed (commented):
nice 😄

cdhanna commented:
I've switched the `DxtBitmap` and `EtcnBitmap`s types to use the Crunch compressor (because it is noticeably faster than basisU). 

I've cleaned up my code a bit so I feel like the integration is workable enough. 

----

>I wonder if we need to separate TextureProcessorOutputFormat.EtcCompressed which will do both ETC1 and ETC2 depending on the Alpha and document the behaviour in the xml comments. And deprecate the Etc1Compressed (somehow).

Maybe a bit too intense, but I've marked `Etc1Compressed` value as `[Obsolete]` and introduced a new `EtcCompressed`. The new value will do both ETC1/2, while the old `Etc1Compressed` does exactly what it used to. I added pragmas to disable the compiler warnings for handling the `Etc1Compressed` value, but perhaps marking it as `[Obsolete]` is too aggressive? 


----

But yeah! Exciting times, because now, _all_ of the original formats are ported over to _basisU_ or _crunch_, and etc2/astc are available.

The important remaining TODO items are to...
1. make official releases of mgcb-basisu, and mgcb-crunch, 
2. figure out how they are installed on monogame-users' projects when folks upgrade to the latest monogame, and/or create a new project. I think this means we need to modify the monogame templates to handle the "fresh project" flow. I'm not sure how best to handle the "upgrade project version" flow. Perhaps it is just a documented upgrade step? We could _automatically_ install the tools if they're not yet installed....
3. I need to figure out how to make a full test release so I can poke around.

cdhanna commented:
And just calling it out again; there are still more unit tests that I cannot run due to library issues with the ARM CPU. 


- `FreeImage` seems to be a problem, 
```
System.DllNotFoundException : Unable to load shared library 'FreeImage' or one of its dependencies. In order to help diagnose loading problems, consider setting the DYLD_PRINT_LIBRARIES environment variable: 
dlopen(/Users/chrishanna/Documents/Github/MonoGame/Artifacts/Tests/Tools/Debug/runtimes/osx/native/FreeImage.dylib, 0x0001):
```


- `Freetype` also,
```
System.DllNotFoundException : Unable to load shared library 'freetype6' or one of its dependencies. In order to help diagnose loading problems, consider setting the DYLD_PRINT_LIBRARIES environment variable: 
dlopen(/Users/chrishanna/Documents/Github/MonoGame/Artifacts/Tests/Tools/Debug/runtimes/osx/native/freetype6.dylib, 0x0001):
```

- I know this one is documented, but calling it out for completion, 
```
Microsoft.Xna.Framework.Content.Pipeline.InvalidContentException : MGFXC effect compiler requires a valid Wine installation to be able to compile shaders.
```

- Assimp ? This gets thrown on the ModelProcessorTests, `MissingNormalsTestCustom`.
```
Assimp.AssimpException : Error loading unmanaged library from path: libassimp.dylib
```

chrisbeamable committed:
restore dxt3

dellis1972 commented:
> And just calling it out again; there are still more unit tests that I cannot run due to library issues with the ARM CPU.
> 
> * `FreeImage` seems to be a problem,
> 
> ```
> System.DllNotFoundException : Unable to load shared library 'FreeImage' or one of its dependencies. In order to help diagnose loading problems, consider setting the DYLD_PRINT_LIBRARIES environment variable: 
> dlopen(/Users/chrishanna/Documents/Github/MonoGame/Artifacts/Tests/Tools/Debug/runtimes/osx/native/FreeImage.dylib, 0x0001):
> ```
> 
> * `Freetype` also,
> 
> ```
> System.DllNotFoundException : Unable to load shared library 'freetype6' or one of its dependencies. In order to help diagnose loading problems, consider setting the DYLD_PRINT_LIBRARIES environment variable: 
> dlopen(/Users/chrishanna/Documents/Github/MonoGame/Artifacts/Tests/Tools/Debug/runtimes/osx/native/freetype6.dylib, 0x0001):
> ```
> 
> * I know this one is documented, but calling it out for completion,
> 
> ```
> Microsoft.Xna.Framework.Content.Pipeline.InvalidContentException : MGFXC effect compiler requires a valid Wine installation to be able to compile shaders.
> ```
> 
> * Assimp ? This gets thrown on the ModelProcessorTests, `MissingNormalsTestCustom`.
> 
> ```
> Assimp.AssimpException : Error loading unmanaged library from path: libassimp.dylib
> ```

All of the FreeType, Assimp things will be handled in another PR.
The Wine Effect compiler still you can setup your dev box using the following instructions https://docs.monogame.net/articles/getting_started/1_setting_up_your_development_environment_unix.html#optional-setup-wine-for-effect-compilation. 


Regarding the Basisu and Crunch Nugets, I'm working. on getting those published. We hit a snag about the nuget files needing a Readme so I'm having to update the buildscripts stuff again.

chrisbeamable committed:
clean up stale comments and unused code.

cdhanna commented:
Oh right, by going through my code, I've remembered that there is still one missing format (my apologies for the false claim earlier). 

The ATC_Explicit_Alpha is not supported by basisU. 
ATC_Interpolated_Alpha _is_ supported. 

I haven't been able to find a wealth of information, except for this, https://registry.khronos.org/OpenGL/extensions/AMD/AMD_compressed_ATC_texture.txt, 

```
    ATC_RGBA_EXPLICIT_ALPHA_AMD Format
    ==================================

    This format compresses blocks of source texels down to 8 bits per texel.
    Assuming 8-bit component source texels, this represents a 4:1 compression
    ratio.  This is generally the best format to use when alpha transitions are
    sharp.

    ATC_RGBA_INTERPOLATED_ALPHA_AMD Format
    ======================================

    This format compresses blocks of source texels down to 8 bits per texel.
    Assuming 8-bit component source texels, this represents a 4:1 compression
    ratio.  This is generally the best format to use when alpha transitions are
    gradient.
```

Sounds like its not ideal but not _horrible_.

dellis1972 commented:
> Oh right, by going through my code, I've remembered that there is still one missing format (my apologies for the false claim earlier).
> 
> The ATC_Explicit_Alpha is not supported by basisU. ATC_Interpolated_Alpha _is_ supported.
> 
> I haven't been able to find a wealth of information, except for this, https://registry.khronos.org/OpenGL/extensions/AMD/AMD_compressed_ATC_texture.txt,
> 
> ```
>     ATC_RGBA_EXPLICIT_ALPHA_AMD Format
>     ==================================
> 
>     This format compresses blocks of source texels down to 8 bits per texel.
>     Assuming 8-bit component source texels, this represents a 4:1 compression
>     ratio.  This is generally the best format to use when alpha transitions are
>     sharp.
> 
>     ATC_RGBA_INTERPOLATED_ALPHA_AMD Format
>     ======================================
> 
>     This format compresses blocks of source texels down to 8 bits per texel.
>     Assuming 8-bit component source texels, this represents a 4:1 compression
>     ratio.  This is generally the best format to use when alpha transitions are
>     gradient.
> ```
> 
> Sounds like its not ideal but not _horrible_.

So I don't know of anyone who uses the `ATCCompressed` option apart from me, and I was the one that put it in lol. A quick github search confirms that, although it doesn't tell us if anyone is using it in private repo's.
But this is likely not very important to support both. If users need a hard alpha they can always use one of the non compressed formats. 

We might be able to revisit this with something along the lines of this https://github.com/Triang3l/S3TConv. Just a thought.

cdhanna commented:
It does appear from the docs that the atc interpolated format _will_ support alpha, but perhaps just not _as_ well as explicit.

dellis1972 commented:
> It does appear from the docs that the atc interpolated format _will_ support alpha, but perhaps just not _as_ well as explicit.

Yup, its gonna be less sharp than explicit

dellis1972 commented:
Actually for now lets compress using the same stuff as fonts if explicit alpha is needed. It won't make any difference as the runtime will be able to use it (see https://github.com/MonoGame/MonoGame/blob/39313ed879c2886997b6641f097ada548e1e6f22/MonoGame.Framework.Content.Pipeline/Graphics/GraphicsUtil.cs#L197).

cdhanna commented:
I'm also looking at this nuget package; https://github.com/Nominom/BCnEncoder.NET/blob/master/BCnEnc.Net/Encoder/AtcBlockEncoder.cs

It appears to do the DXT encodings in-process; which _may_ represent a significant performance boost for the DXT sub-types. Digging through the code, it also appears to have some references to ATC interpolated and explicit, so I am _curious_...

cdhanna commented:
>I'm also looking at this nuget package; https://github.com/Nominom/BCnEncoder.NET/blob/master/BCnEnc.Net/Encoder/AtcBlockEncoder.cs

I need to switch gears right now, but from some initial tests, it looks like this may be an order-of-magnitude faster than invoking the `crunch` process. The BcnEncoder library doesn't cover all the formats, so we'd still need crunch for ETC1/2. 
But from a quick test running the DXT1 compression, invoking crunch causes the test to pass in 800ms, but the nuget package passes the test in 80ms.

chrisbeamable committed:
use bcn encoder library to support atc explicit alpha

cdhanna commented:
Okay, so that library supports ATC interpolated alpha and explicit alpha. It also supports DXT1/a, and DXT5. It _theoretically_ supports DXT3, but I was hitting a snag with the premultiplied alpha, so I left DXT3 compression using Crunch. 

The speeds are way faster when all the compression can happen in-process with no file input/output.

dellis1972 commented:
@cdhanna 

Packages are available now 
https://www.nuget.org/packages/mgcb-basisu/ 
https://www.nuget.org/packages/mgcb-crunch/

chrisbeamable committed:
update crunch to deployed version

cdhanna commented:
Thanks, I've updated the versions in the `.config/dotnet-tools.json` file for local development. 

I think it is time to open the official PR, and address whatever comments come through there. 
To be honest, I'm still unclear exactly how the tool resolution code is meant to work, but I don't think that should block the opening of the PR.

chrisbeamable committed:
Merge branch 'develop' into basisu

dellis1972 commented:
> Thanks, I've updated the versions in the `.config/dotnet-tools.json` file for local development.
> 
> I think it is time to open the official PR, and address whatever comments come through there. To be honest, I'm still unclear exactly how the tool resolution code is meant to work, but I don't think that should block the opening of the PR.

I agree, thanks for all the great work 😄

chrisbeamable committed:
remove unused namespaces

cdhanna commented:
Closing this PR so discussion can continue in one place, on the main MonoGame repo, 
https://github.com/MonoGame/MonoGame/pull/8456
